R/collapse_intron_counts.R

```diff
@@ -53,8 +53,16 @@ collapse_intron_counts <- function(counts,
   }
   # remove -I -U -A at the end of the gene names
   genes <- str_remove(rownames(counts), "-[IUA]$")
-  # aggregate Matrix counts by gene name
-  counts <- DelayedArray::rowsum(counts, genes) |> as("dgCMatrix")
+
+  # aggregate Matrix counts by gene name using matrix math
+  # get unique gene names, encode genes as factors
+  ugenes <- unique(genes)
+  fgenes <- factor(genes, levels = ugenes)
+  # create a transformer matrix, with 1s at positions that correspond to old/new names, zeros elsewhere
+  transformer <- Matrix::sparseMatrix(as.integer(fgenes), 1:length(fgenes), x =1)
+  rownames(transformer) <- ugenes
```

Comment:
I'm a little confused on how these dimensions match up? The number of rows is going to be all the genes, while the rownames will be only the unique genes?

Reply:
The number of rows is the number of genes, the number of columns is gene parts (gene-U,-A, etc.). 

What I am doing here is making a sparse matrix with values at specific locations, corresponding to a factor level that is the unique gene (from `as.integer(fgenes)` and position in the gene part list (from `1:length(fgenes)`), which corresponds to the rows of the original counts matrix. 

So if there are 600 gene parts from 200 genes, this will be a 200x600 matrix where each column has a single value with a value of 1, and each row will have 3, corresponding to the three parts of that gene. Then we multiply that transformer by our 600xN_cells matrix to get a 200xN_cells matrix at the end.

Reply:
I only fill in the row names, as those will be preserved in the output. The gene part names (columns) will go away, so I didn't bother adding those in.

Reply:
To be a bit more explicit... `sparseMatrix(c(1,1,2,2), c(1,2,3,4), x=1)` will give something like:
```
     [,1] [,2] [,3] [,4]
[1,]    1    1    0    0
[2,]    0    0    1    1
```
(but sparse) 

So I am only indirectly specifying the dimensions by the max value of the levels of the `fgenes` object for the number of rows, and its length for the number of columns.

Reply:
Ahh this makes so much more sense now. Thanks for the example! I was thinking that it was making a matrix with the dimensions equivalent to the number of gene parts x gene parts, which is why I was getting tripped up.